src/foam/util/AddressUtil.java

```diff
@@ -57,4 +66,73 @@
     return new String[] { streetNumber, streetName };
   }
 
+  public static String normalizeRegion(X x, String country, String regionCode) {
+    String normalizedRegion = country + "-" + regionCode;
+    DAO regionDAO = (DAO) x.get("regionDAO");
+    Region found = (Region) regionDAO.find(regionCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) regionDAO.where(AND(
+      EQ(Region.COUNTRY_ID, country),
+      OR(
+        EQ(Region.ISO_CODE, regionCode),
+        CONTAINS_IC(Region.NAME, regionCode)
+      )
+    )).select(new ArraySink());
+
+    List<Region> regions = sink.getArray();
+    
+    // Take the first match as the default
+    if (regions.size() > 0)
+      normalizedRegion = regions.get(0).getCode();
+    
+    // When there are more than one, take the first matches on the ISO code or on the full name (i.e. not just CONTAINS_IC)
+    if (regions.size() > 1) {
+      for ( int i = 0; i < regions.size(); i++ ) {
+        Region region = (Region) regions.get(i);
+        if ( region.getIsoCode().equals(regionCode) || region.getName().equalsIgnoreCase(regionCode) ) {
+          normalizedRegion = region.getCode();
+          break;
+        }
+      }
+    }
+    
+    return normalizedRegion;
+  }
+
+  public static String normalizeCountry(X x, String countryCode) {
+    String normalizedCountry = countryCode;
+    DAO countryDAO = (DAO) x.get("countryDAO");
+    Country found = (Country) countryDAO.find(countryCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) countryDAO
+      .where(OR(
+        EQ(Country.ISO31661CODE, countryCode),
+        CONTAINS_IC(Country.NAME, countryCode)
```

Comment:
country code should never be partial .. it should either be the whole country name, or either 3 digit ISO code or the 2 digit ISO code, so we should never get "United", only "United States" or "United Kingdom". The problem comes when we get something in IND and it matches Indonesia. In this case the logic below would kick in and the exact match on 3 digit ISO code would properly return India (IN) instead of Indonesia (ID)

Reply:
assuming we never get "United" but if we do, should we throw instead of return US or UK?

Reply:
I think a best effort match is better than entering an invalid country or throwing and not being able to save the address at all ...

Reply:
Ok if we decided so, can we add documentation for future reference?